File: ZeroTier One/ShowNetworksViewController.m
//
//  ShowNetworksViewController.m
//  ZeroTier One
//
//  Created by Grant Limberg on 8/7/16.
//  Copyright © 2016 ZeroTier, Inc. All rights reserved.
//

#import "ShowNetworksViewController.h"
#import "NetworkMonitor.h"
#import "NetworkInfoCell.h"
#import "Network.h"

@interface ShowNetworksViewController ()

@end

@implementation ShowNetworksViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    [self.tableView setDelegate:self];
    [self.tableView setDataSource:self];
    [self.tableView setBackgroundColor:[NSColor clearColor]];
}

- (void)viewWillAppear {
    [super viewWillAppear];
    self.visible = YES;
}

- (void)viewWillDisappear {
    [super viewWillDisappear];
    self.visible = NO;
}

- (void)deleteNetworkFromList:(NSString *)nwid {
    [self.netMonitor deleteSavedNetwork:nwid];
}

- (void)setNetworks:(NSArray<Network *> *)list {
    _networkList = list;
    if(_visible) {
        [_tableView reloadData];
    }
}

- (NSInteger)numberOfRowsInTableView:(NSTableView *)tableView {
    return [_networkList count];
}

- (NSView*)tableView:(NSTableView *)tableView viewForTableColumn:(NSTableColumn *)tableColumn row:(NSInteger)row
{
    NetworkInfoCell *cell = (NetworkInfoCell*)[tableView makeViewWithIdentifier:@"NetworkInfoCell"
                                                                          owner:nil];
    Network *network = [_networkList objectAtIndex:row];
    cell.parent = self;
    cell.networkIdField.stringValue = [NSString stringWithFormat:@"%10llx", network.nwid];
    cell.networkNameField.stringValue = network.name;
    cell.statusField.stringValue = [network statusString];
    cell.typeField.stringValue = [network typeString];
    cell.mtuField.stringValue = [NSString stringWithFormat:@"%d", network.mtu];
    cell.macField.stringValue = network.mac;
    cell.broadcastField.stringValue = network.broadcastEnabled ? @"ENABLED" : @"DISABLED";
    cell.bridgingField.stringValue = network.bridge ? @"ENABLED" : @"DISABLED";
    cell.deviceField.stringValue = network.portDeviceName;

    if(network.connected) {
        cell.connectedCheckbox.state = NSOnState;

        if(network.allowDefault) {
            cell.allowDefault.enabled = YES;
            cell.allowDefault.state = NSOnState;
        }
        else {
            cell.allowDefault.state = NSOffState;

            if([Network defaultRouteExists:_networkList]) {
                cell.allowDefault.enabled = NO;
            }
            else {
                cell.allowDefault.enabled = YES;
            }
        }

        cell.allowGlobal.enabled = YES;
        cell.allowManaged.enabled = YES;
    }
    else {
        cell.connectedCheckbox.state = NSOffState;
        cell.allowDefault.enabled = NO;
        cell.allowGlobal.enabled = NO;
        cell.allowManaged.enabled = NO;
    }

    cell.allowGlobal.state = network.allowGlobal ? NSOnState : NSOffState;
    cell.allowManaged.state = network.allowManaged ? NSOnState : NSOffState;

    cell.addressesField.stringValue = @"";

    for(NSString *addr in network.assignedAddresses) {
        cell.addressesField.stringValue = [[cell.addressesField.stringValue stringByAppendingString:addr] stringByAppendingString:@"\n"];
    }

    return cell;
}

@end

File: ZeroTier One/ShowNetworksViewController.swift
//
//  ShowNetworksViewController.swift
//  ZeroTier One
//
//  Created by Grant Limberg on 5/14/16.
//  Copyright © 2016 ZeroTier, Inc. All rights reserved.
//

import Cocoa

class ShowNetworksViewController: NSViewController, NSTableViewDelegate, NSTableViewDataSource {

    @IBOutlet var tableView: NSTableView?

    var networkList: [Network] = [Network]()
    var netMonitor: NetworkMonitor!

    var visible = false

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do view setup here.

        tableView?.setDelegate(self)
        tableView?.setDataSource(self)
        tableView?.backgroundColor = NSColor.clearColor()
    }

    private func  dataFile() -> String {
        var appSupport = NSFileManager.defaultManager().URLsForDirectory(.ApplicationSupportDirectory, inDomains: .UserDomainMask)[0]
        appSupport = appSupport.URLByAppendingPathComponent("ZeroTier").URLByAppendingPathComponent("One").URLByAppendingPathComponent("networks.dat")
        return appSupport.path!
    }

    override func viewWillAppear() {
        super.viewWillAppear()
        visible = true
    }

    override func viewWillDisappear() {
        super.viewWillDisappear()

        let filePath = dataFile()
        NSKeyedArchiver.archiveRootObject(self.networkList, toFile: filePath)
        visible = false
    }

    func deleteNetworkFromList(nwid: String) {
        netMonitor.deleteSavedNetwork(nwid)
    }


    func setNetworks(list: [Network]) {
        networkList = list
        if visible {
            tableView?.reloadData()
        }
    }

    // NSTableViewDataSource

    func numberOfRowsInTableView(tableView: NSTableView) -> Int {
        return networkList.count
    }

    // end NSTableViewDataSource

    // NSTableViewDelegate

    func tableView(tableView: NSTableView, viewForTableColumn tableColumn: NSTableColumn?, row: Int) -> NSView? {
        if let cell = tableView.makeViewWithIdentifier("NetworkInfoCell", owner: nil) as? NetworkInfoCell {
            let network = networkList[row]
            cell.parent = self
            cell.networkIdField.stringValue = String(network.nwid, radix: 16)
            cell.networkNameField.stringValue = network.name

            cell.statusField.stringValue = network.status.description
            cell.typeField.stringValue = network.type.description
            cell.macField.stringValue = network.mac
            cell.mtuField.stringValue = String(network.mtu)
            cell.broadcastField.stringValue = network.broadcastEnabled ? "ENABLED" : "DISABLED"
            cell.bridgingField.stringValue = network.bridge ? "ENABLED" : "DISABLED"
            cell.deviceField.stringValue = network.portDeviceName

            if network.connected {
                cell.connectedCheckbox.state = NSOnState

                if network.allowDefault {
                    cell.allowDefault.enabled = true
                    cell.allowDefault.state = NSOnState
                }
                else {
                    cell.allowDefault.state = NSOffState

                    if defaultRouteExists(networkList) {
                        cell.allowDefault.enabled = false
                    }
                    else {
                        cell.allowDefault.enabled = true
                    }
                }

                cell.allowGlobal.enabled = true
                cell.allowManaged.enabled = true
            }
            else {
                cell.connectedCheckbox.state = NSOffState

                cell.allowDefault.enabled = false
                cell.allowGlobal.enabled = false
                cell.allowManaged.enabled = false
            }
            
            cell.allowGlobal.state = (network.allowGlobal ? NSOnState : NSOffState)
            cell.allowManaged.state = (network.allowManaged ? NSOnState : NSOffState)

            cell.addressesField.stringValue = ""

            for nw in network.assignedAddresses {
                cell.addressesField.stringValue += "\(nw)\n"
            }



            return cell
        }

        return nil
    }

    // end NSTableViewDelegate
}
